Add batch vote summary route

martonp opened the pull request:
Add batch vote summary route which fetches a summary of the voting process for a set of proposals.

- Add BatchVoteSummary command to decredplugin which gets `VoteSummaryReply` for multiple proposals.
- Update documentation to say VoteStatus routes are deprecated.
- Update cache that stores `VoteStatus` for proposals with completed voting to store `VoteSummary`, and convert between the two in the deprecated routes.

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +887,69 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+
+// processBatchVoteSummary the voting summary information for a set of
+// proposals.
+func (p *politeiawww) processBatchVoteSummary(batchVoteSummary www.BatchVoteSummary) (*www.BatchVoteSummaryReply, error) {
```
Can we limit the max number of vote summaries being requested to the proposal page size and return a user error if this limit is exceeded?

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +887,69 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+		}
+	}
+
+	reply := www.BatchVoteSummaryReply{
```
There's not really any point in creating a variable here.  Can we just do this instead:

```
	return &www.BatchVoteSummaryReply{
		Summaries: summaries,
	}, nil
```

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +887,69 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+	}
+
+	if len(summaries) != len(batchVoteSummary.Tokens) {
+		return nil, www.UserError{
```
A user error contains a `ErrorContext` field that allows you to give context to the error message when appropriate.  It would make sense here to use the `ErrorContext` to specify which tokens were not found.

lukebp commented on politeiawww/api/www/v1/v1.go:
```diff
@@ -750,6 +763,17 @@ type BatchProposalsReply struct {
+}
+
+// BatchVoteSummaryReply is used to reply to a BatchVoteSummary command.
+type BatchVoteSummaryReply struct {
```
Would it make more sense to remove the `BestBlock` from the `VoteSummary` and add it to the `BatchVoteSummaryReply` instead?

martonp replied:
Yeah it does

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +887,69 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+// proposals.
+func (p *politeiawww) processBatchVoteSummary(batchVoteSummary www.BatchVoteSummary) (*www.BatchVoteSummaryReply, error) {
+	log.Tracef("processBatchVoteSummary")
+
```
Can we validate the passed in tokens first and return a user error one or more of the tokens is invalid?  The function `tokenIsValid()` can be used to do this.

martonp commented:
@lukebp this is ready to review

lukebp commented on politeiawww/proposals.go:
```diff
@@ -152,6 +152,18 @@ func tokenIsValid(token string) bool {
 }
 
+func getInvalidTokens(tokens []string) []string {
+	invalidTokens := make([]string, 0)
```
```suggestion
	invalidTokens := make([]string, 0, len(token))
```

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +899,89 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+
+// processBatchVoteSummary the voting summary information for a set of
+// proposals.
+func (p *politeiawww) processBatchVoteSummary(batchVoteSummary www.BatchVoteSummary) (*www.BatchVoteSummaryReply, error) {
```
If you look at the `processVoteStatus` function you'll notice that we memoize the vote results for proposals that have finished voting in a memory cache in order to improve performance.  We need to do the same thing here.

The problem you're going to run into is that the `VoteResult` and `VoteStatusReply` don't map 1-to-1.  The way I would do this would be to convert the `p.voteStatusReply` memory cache to hold vote results instead then update the vote status endpoint to cache a `VoteResult` instead of a `VoteStatusReply`.  This way, when we go to remove the deprecated vote status endpoints we don't have to worry about updating the memory cache.  We can just delete the endpoints.

martonp replied:
I don't see any `VoteResult` type anywhere. I can cache `decredplugin.VoteSummaryReply` though.

martonp replied:
That's kind of strange too though... i think it would be the cleanest to just cache both the`VoteSummary` and the `VoteStatusReply` and then remove the `VoteStatusReply` cache when deprecating that route, no?

lukebp replied:
> I don't see any VoteResult type anywhere

Typo.  I meant `VoteSummary`.

> i think it would be the cleanest to just cache both theVoteSummary and the VoteStatusReply and then remove the VoteStatusReply cache when deprecating that route, no?

I'm really not a fan of having two caches with the same data in them.  I think its cleaner to have just a `VoteSummary` cache, since the vote status routes are being deprecated, and then to have helper functions like the ones below for vote status routes until they're fully removed.

```
func (p *politeiawww) cacheVoteStatusReply(v www.VoteStatusReply) {
	// Convert VoteStatusReply to VoteSummary
	// Cache VoteSummary
}

func (p *politeiawww) getVoteStatusReply(token string) (*www.VoteStatusReply, bool) {
	// Lookup VoteSummary for token
	// Convert VoteSummary to VoteStatusReply
}
```

martonp commented:
@lukebp  ready to go.

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +899,126 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+		return voteSummaries, nil
+	}
+
+	r, err := p.decredBatchVoteSummary(tokens)
```
Shouldn't this just be looking up `tokensToLookup`?

lukebp commented on politeiawww/proposals.go:
```diff
@@ -890,6 +899,126 @@ func (p *politeiawww) processProposalDetails(propDetails www.ProposalsDetails, u
+
+		endHeight, err := strconv.ParseUint(summary.EndHeight, 10, 64)
+		if err != nil {
+			endHeight = 0
```
We should probably do a `log.Errorf` here to alert us that something is amiss with the cache data.

lukebp commented on politeiawww/proposals.go:
```diff
@@ -1504,25 +1633,78 @@ func voteResults(sv www.StartVote, cv []www.CastVote) []www.VoteOptionResult {
+		return nil, false
+	}
+
+	totalVotes := uint64(0)
```
Style nit. If you're initializing a variable to its zero value just do:

```suggestion
	var totalVotes uint64
```

martonp commented:
@lukebp good to go

martonp commented:
I've updated the first comment of the PR to be more descriptive, and I've updated the title of the PR (which is what's used in the squashed commit message?). Does this meet the expectation?
